cvmfs/voms_authz/voms_authz.cc

```diff
@@ -0,0 +1,406 @@
+}
+
+class VOMSSessionCache {
+
+public:
+
+    VOMSSessionCache()
+      : m_zombie(true),
+        m_last_clean(get_mono_time())
+    {
+        pthread_mutex_init(&m_mutex, NULL);
+        load_voms_library();
+    }
+
+    ~VOMSSessionCache()
+    {
+        for (KeyToVOMS::const_iterator it = m_map.begin(); it!=m_map.end(); it++)
+        {
+            (*g_VOMS_Destroy)(it->second.first);
+        }
+        close_voms_library();
+        pthread_mutex_destroy(&m_mutex);
+    }
+
+    bool get(pid_t pid, struct vomsdata *&vomsinfo)
+    {
+        if (m_zombie) {return false;}
+        time_t now = get_mono_time();
+        KeyType mykey;
+        if (!lookup(pid, mykey)) {return false;}
+        LogCvmfs(kLogVoms, kLogDebug, "PID %d maps to session %d, bday %llu", pid, mykey.first, mykey.second);
+        pthread_mutex_lock(&m_mutex);
+        if (now > m_last_clean + 100) {clean_tables();};
+        KeyToVOMS::const_iterator iter = m_map.find(mykey);
+        pthread_mutex_unlock(&m_mutex);
+        if (iter == m_map.end())
+        {
+            return false;
+        }
+        LogCvmfs(kLogVoms, kLogDebug, "Session %d, bday %llu has cached VOMS data", mykey.first, mykey.second);
+        vomsinfo = iter->second.first;
+        return true;
+    }
+
+    // Put the VOMS data into the map.  Map owns the pointer
+    // after this returns.  This function is thread safe.
+    //
+    // The returned pointer is the data in the map which is possibly
+    // not the input vomsdata.
+    //
+    // If result is NULL, then the operation failed and errno is set.
+    struct vomsdata *try_put(pid_t pid, struct vomsdata *voms_ptr)
+    {
+        if (m_zombie) {return false;}
+        KeyType mykey;
+        if (!lookup(pid, mykey))
+        {
+            LogCvmfs(kLogVoms, kLogDebug, "Failed to determine session key for PID %d.", pid);
+            return false;
+        }
+        time_t now = get_mono_time();
+
+        pthread_mutex_lock(&m_mutex);
+        std::pair<KeyToVOMS::iterator, bool> result = m_map.insert(std::make_pair(mykey, std::make_pair(voms_ptr, now)));
+        pthread_mutex_unlock(&m_mutex);
+        if (!result.second)
+        {
+            (*g_VOMS_Destroy)(voms_ptr);
+        }
+        LogCvmfs(kLogVoms, kLogDebug, "Cached VOMS data for session %d, bday %llu.", mykey.first, mykey.second);
+        return result.first->second.first;
+    }
+
+private:
+
+    typedef std::pair<pid_t, unsigned long long> KeyType;
+    typedef std::pair<struct vomsdata*, time_t> ValueType;
+    typedef std::map<KeyType, ValueType> KeyToVOMS;
+
+    typedef std::map<KeyType, KeyType> PidToSid;
+    PidToSid m_pid_map;
+
+    // Lookup pid's sid and birthdy.
+    // Returns false on failure and sets errno.
+    bool lookup(pid_t pid, KeyType &mykey)
+    {
+        char pidpath[PATH_MAX];
+        if (snprintf(pidpath, PATH_MAX, "/proc/%d/stat", pid) >= PATH_MAX)
```

Comment:
I don't know much about how proc handles inodes, that's a good idea to check.

However, since proc is not a "real filesystem", I'm not sure how much of a performance hit/hindrance we're looking at by reading out this file.

I know this is about equivalent to the amount of work done when invoking getrusage (glibc's internal implementation opens and parses this file), and we invoke that quite a bit within CMSSW...

It'll be worth measuring.